synapse/cortex.py

```diff
@@ -4848,6 +4848,60 @@ async def addLayer(self, ldef=None, nexs=True):
         else:
             return await self._addLayer(ldef, (None, None))
 
+    async def _twinLayer(self, oldlayr):
+
+        newldef = s_msgpack.deepcopy(oldlayr.layrinfo)
+
+        newldef.pop('iden', None)
+
+        newldef = await self.addLayer(newldef)
+        newlayr = self.reqLayer(newldef.get('iden'))
+
+        oldinfo = self.auth.reqAuthGate(oldlayr.iden).pack()
+
+        for userinfo in oldinfo.get('users', ()):
+
+            user = self.auth.user(userinfo.get('iden'))
+            if user is None: # pragma: no cover
+                continue
+
+            if userinfo.get('admin'):
+                await user.setAdmin(True, gateiden=newlayr.iden)
+
+            for rule in userinfo.get('rules', ()):
+                await user.addRule(rule, gateiden=newlayr.iden)
+
+        for roleinfo in oldinfo.get('roles', ()):
+
+            role = self.auth.role(roleinfo.get('iden'))
+            if role is None: # pragma: no cover
+                continue
+
+            for rule in roleinfo.get('rules', ()):
+                await role.addRule(rule, gateiden=newlayr.iden)
+
+        return newlayr
+
+    @s_nexus.Pusher.onPushAuto('layer:swap')
+    async def swapLayer(self, oldiden, newiden):
+        '''
+        Atomically swap out a layer from all views that contain it.
+        '''
+        self.reqLayer(oldiden)
+        self.reqLayer(newiden)
+
+        for view in self.views.values():
```

Comment:
Need to make a copy of this to avoid it changing size during iteration

Reply:
I think it's safe within a nexus handler, but it's essentially free so i'll add the `list()` call :)